### ISSUE TITLE
Reward Shaping Experiment — 對漏接球／Life Loss 加入負 Reward，並比較實際模型成果

### ISSUE THREAD

**Issue by Tommyweige** [Tommyweige/breakout-rl-engineering#9]
## 目標

目前 Breakout 訓練主要依賴遊戲原本的 reward：打到磚塊時得到正 reward，沒有得分時 reward 為 0；在目前 `reward_clip=true` 的訓練設定下，正 reward 會被轉成 `+1`。

這個 Issue 要做一個新的 **reward shaping 實驗**：

> 當 Agent 沒有接到球、因此掉一條命（life loss）時，額外給模型一個負 reward。

希望回答一個很直接的問題：

> **除了「打到磚塊會加分」之外，再明確告訴模型「漏接球是不好的」，能不能讓模型學得更快、活得更久，最後拿到更高的 Breakout 分數？**

這不是直接覆蓋目前 final model，而是一個可重現、可比較的實驗。只有在結果真的比較好時，才考慮是否取代原本的 reward 設計。

---

## 現況

目前環境已經具備 life loss 偵測。

`BreakoutFireResetWrapper` 在每次 step 後會比較 lives，並透過：

```text
info["fire_reset_life_loss"]
```

標記這一步是否發生掉命。

因此這個實驗 **不需要從畫面自行判斷球有沒有掉下去**，也不需要新增 CV / ball tracking。

目前訓練 reward 的概念可以視為：

```text
打到磚塊 -> +1
其他情況 -> 0
```

在原始 Atari reward 經過 reward clipping 後，正 reward 會變成 `+1`。

目前沒有額外的：

```text
漏接球 / 掉命 -> negative reward
```

---

# Experiment Hypothesis

新的 reward 設計：

```text
打到磚塊                +1
沒有得分                  0
漏接球 / life loss        -1
```

第一版固定使用：

```text
life_loss_penalty = -1.0
```

核心假設：

1. 原始 reward 只告訴 Agent「打到磚塊很好」。
2. 但當球掉下去時，如果 reward 只是 `0`，模型只會間接透過「之後拿不到分」去理解這是不好的。
3. 加入明確的 `-1`，可能讓 Agent 更快學會：
   - paddle 應該靠近球落點
   - 不要做容易造成漏接的動作
   - 延長單條生命的存活時間
4. 但 reward shaping 也可能帶來副作用：
   - 模型變得過度保守
   - Q-value scale 改變
   - 為了避免 -1 而犧牲更積極的打磚塊策略
   - 最終 raw game score 反而下降

所以這個 Issue 的重點不是「一定要讓 -1 版本獲勝」，而是要做出可以相信的比較結果。

---

# Important Rule：遊戲分數與訓練 Reward 必須分開

**不要修改 Atari / ALE 原本的遊戲 reward。**

必須同時保留兩種數值：

### 1. Raw game reward

來自：

```python
observation, raw_reward, terminated, truncated, info = env.step(action)
```

用途：

- 計算真正 Breakout 分數
- evaluation
- 和現有模型公平比較

### 2. Training reward

模型實際存進 replay buffer、用來計算 Bellman target 的 reward。

新的邏輯：

```python
training_reward = clipped_game_reward

if info.get("fire_reset_life_loss", False):
    training_reward += life_loss_penalty
```

第一版：

```python
life_loss_penalty = -1.0
```

也就是：

```text
raw reward > 0                    -> +1
raw reward = 0                    -> 0
life loss                         -> 額外 -1
```

### 注意 reward clipping 的順序

不要寫成：

```python
sign(raw_reward + life_loss_penalty)
```

因為這會讓 penalty magnitude 被 sign clipping 吃掉，也會把遊戲 reward 與 shaping 混在一起。

應該使用：

```python
clipped_game_reward = sign(raw_reward)
training_reward = clipped_game_reward + life_loss_penalty_if_needed
```

這樣未來如果要測：

```text
-0.25
-0.5
-1.0
-2.0
```

才有意義。

---

# Implementation

## 1. 新增 reward shaping configuration

在 training config 加入明確設定，例如：

```json
{
  "reward_clip": true,
  "life_loss_penalty": 0.0
}
```

baseline 必須維持：

```json
"life_loss_penalty": 0.0
```

新的實驗 config 使用：

```json
"life_loss_penalty": -1.0
```

不要把 `-1` hard-code 在 trainer 裡。

建議新增獨立 config，例如：

```text
configs/dueling_double_dqn_life_loss_penalty.json
```

並確保其他超參數和比較基準相同。

---

## 2. 實作 reusable reward shaping function

不要把判斷散落在 training loop。

建議新增明確 function，例如：

```python
def shape_training_reward(
    raw_reward: float,
    *,
    reward_clip: bool,
    life_loss: bool,
    life_loss_penalty: float,
) -> float:
    ...
```

預期行為：

```text
raw=0, life_loss=False, penalty=-1 -> 0
raw=1, life_loss=False, penalty=-1 -> +1
raw=0, life_loss=True,  penalty=-1 -> -1
raw=1, life_loss=True,  penalty=-1 -> 0
```

function 必須處理：

- finite validation
- `life_loss_penalty <= 0`
- reward clipping on/off
- backward compatibility：penalty=0 時結果必須與目前 trainer 完全相同

---

## 3. Trainer 接入 life-loss penalty

目前 environment 已經會回傳：

```python
info["fire_reset_life_loss"]
```

單環境 trainer 與 vectorized trainer 都必須支援相同 reward shaping。

流程應為：

```text
env.step(action)
    ↓
raw_reward
    ↓
確認這一步是否 life loss
    ↓
建立 training_reward
    ↓
training_reward 寫進 Replay Buffer
    ↓
raw_reward 繼續用於 episode game score
```

**不要用 shaped reward 當最終遊戲分數。**

---

## 4. Metrics：把兩種 reward 都記錄下來

至少新增／保存：

```text
raw_episode_return
training_episode_return
life_loss_count
life_loss_penalty_total
```

每個 training run 最後要能回答：

```text
這一局真正遊戲拿幾分？
模型學習時總共收到多少 shaped reward？
這一局掉了幾條命？
```

建議 training curve 同時畫：

1. raw episode return
2. shaped episode return
3. life losses per episode
4. rolling raw score
5. rolling episode length

---

# Experiment Design

## Experiment A — Baseline

reward：

```text
brick / positive Atari reward -> +1
life loss                    -> 0 additional penalty
```

設定：

```text
life_loss_penalty = 0.0
```

## Experiment B — Life-loss penalty

reward：

```text
brick / positive Atari reward -> +1
life loss                    -> -1 additional penalty
```

設定：

```text
life_loss_penalty = -1.0
```

---

# Fair Comparison

除了 reward shaping 以外，A / B 必須盡可能保持一致：

```text
architecture
algorithm
training seed
training transitions / steps
gamma
learning rate
batch size
replay size
learning starts
target update interval
epsilon schedule
frame skip
frame stack
sticky action probability
FIRE reset
optimizer
device / precision
```

使用目前專案的最終主力架構：

```text
Dueling Double DQN
```

不要同時更改 architecture、epsilon、learning rate 或其他 hyperparameter，否則無法知道改善是不是來自 reward shaping。

---

# Training Plan

## Stage 1 — Smoke Test

先跑短訓練，確認：

- life loss 能正確偵測
- replay buffer 真的會收到負 reward
- raw score 不受 shaping 影響
- penalty=0 完全維持舊行為
- vectorized trainer 正常
- CUDA path 正常

Smoke test 不拿來判定模型優劣。

---

## Stage 2 — Full Training

至少完成一個與目前 final-model 訓練 budget 同級的完整 shaped-reward run。

第一個正式實驗：

```text
Algorithm: Dueling Double DQN
life_loss_penalty: -1.0
Training seed: 與目前 final model 相同
Training budget: 與目前 final model 相同
```

保存：

```text
checkpoint
final model
training metrics
training curves
config
runtime metadata
reward-shaping metadata
```

---

# Evaluation

最重要的一點：

> **Evaluation 仍然只看原本 Atari raw game score。**

不要把 `-1` life-loss penalty 算進最終 Breakout 成績。

## Primary comparison

使用完全相同的 evaluation contract 與相同 seed set，比較：

```text
Current final model
vs
Life-loss-penalty model
```

至少使用目前 Day 30 已建立的：

```text
50 unique evaluation seeds
```

對每一個 seed 做 paired comparison。

輸出：

| Metric | Baseline | Life-loss penalty | Delta |
|---|---:|---:|---:|
| Mean raw score | | | |
| Median raw score | | | |
| Std | | | |
| P10 | | | |
| P90 | | | |
| Min | | | |
| Max | | | |
| Mean lives lost | | | |
| Mean episode length | | | |
| 50-seed wins | | | |
| 50-seed ties | | | |
| 50-seed losses | | | |

另外計算每個 seed：

```text
score_delta = shaped_model_score - baseline_model_score
```

並畫出 paired score difference。

---

# Learning Efficiency

除了 final score，也要看它是不是「比較快學會」。

比較兩個 training run 在相同 transition 數下的：

```text
raw score rolling mean
raw score rolling median
life loss frequency
average survival / episode length
Q-value statistics
TD error
```

至少截取：

```text
25%
50%
75%
100%
```

training budget 的 checkpoint 或 evaluation point。

要回答：

> 加入負 reward 後，是不是更早開始學會接球？

而不只看最後一次 checkpoint。

---

# Visualizations

至少產出以下圖表：

### 1. Training raw-score curve

```text
Baseline vs Life-loss penalty
```

### 2. Life losses over training

觀察模型是否真的更少漏接。

### 3. 50-seed final score comparison

paired scatter / connected-pair plot 或其他清楚形式。

### 4. Score delta distribution

```text
new_score - baseline_score
```

### 5. Evaluation summary

Mean / Median / P10 / P90 / Min / Max。

圖表與數值都保存成 artifacts，不只截圖。

---

# Optional Follow-up：Penalty Sweep

如果 `-1.0` 確認可以正常訓練，但結果沒有明顯勝負，可再測：

```text
0.0
-0.25
-0.5
-1.0
-2.0
```

但 **這不是第一階段的必要條件**。

第一階段先回答最單純的：

```text
原版 0
vs
life loss -1
```

避免一次把實驗做得太複雜。

---

# Tests

至少新增測試覆蓋：

- [ ] penalty=0 時和現有 `_training_reward` 行為一致
- [ ] raw=0 + no life loss -> 0
- [ ] raw=1 + no life loss -> +1
- [ ] raw=0 + life loss + penalty=-1 -> -1
- [ ] raw=1 + life loss + penalty=-1 -> 0
- [ ] reward_clip=false 時保留 raw reward magnitude
- [ ] invalid / NaN / inf penalty fail closed
- [ ] positive life-loss penalty 不允許
- [ ] single-env trainer 正確寫入 shaped reward
- [ ] vectorized trainer 正確處理每個 env 的 life-loss mask
- [ ] raw episode return 不被 shaped reward 汙染
- [ ] evaluation score 不使用 shaped reward

---

# Artifacts

建議新增獨立目錄，例如：

```text
assets/reward-shaping-life-loss/
```

至少包含：

```text
experiment-config.json
baseline-summary.json
life-loss-penalty-summary.json
paired-evaluation.json
training-comparison.json
training-score-curve.png
life-loss-curve.png
paired-score-comparison.png
score-delta-distribution.png
```

模型本身放在既有 model / checkpoint 結構中即可，不要複製巨大檔案到不必要的位置。

---

# Reader-facing Note

如果這個實驗最後整理成文章／文件，請用初學者能理解的方式解釋：

原本的模型大概只收到：

```text
打到磚塊：做得好，+1
其他時候：沒有特別反應，0
```

新的版本多了一句：

```text
球掉下去：這不好，-1
```

接著再說明：

> Reward shaping 並不保證一定比較好。它只是把我們希望模型重視的事情更明確地寫進學習訊號裡，最後仍然要用真正的遊戲分數驗證。

不要把 shaped reward 當作模型真的多拿到的 Breakout 分數。

---

# Acceptance Criteria

- [ ] `life_loss_penalty` 成為可設定 training config
- [ ] default / baseline 為 `0.0`
- [ ] `-1.0` reward-shaping config 可執行
- [ ] life loss 使用環境既有 `fire_reset_life_loss` 訊號，不使用影像 heuristic
- [ ] raw game reward 與 training reward 完全分離
- [ ] replay buffer 使用 shaped training reward
- [ ] episode/evaluation raw score 仍使用 Atari 原始 reward
- [ ] single-env trainer 支援
- [ ] vectorized trainer 支援
- [ ] 對 reward shaping 有完整 unit tests
- [ ] baseline backward compatibility 通過
- [ ] 完成 smoke training
- [ ] 完成至少一個 full `life_loss_penalty=-1.0` training run
- [ ] 使用相同 Dueling Double DQN architecture
- [ ] 除 reward shaping 外不任意修改核心 hyperparameters
- [ ] baseline 與 shaped model 使用同一 evaluation contract
- [ ] 在相同 50 unique seeds 上完成 paired evaluation
- [ ] 輸出 mean / median / std / P10 / P90 / min / max
- [ ] 輸出 win / tie / loss 與 per-seed score delta
- [ ] 比較 life-loss frequency / survival 或 episode length
- [ ] 產出 training/evaluation visualizations
- [ ] artifacts 可重現並保存完整 config / metadata
- [ ] 最後明確回答 reward shaping 是否改善模型，而不是只說「成功加入 -1 reward」

---

# Definition of Done

這個 Issue 完成後，我們應該可以用實驗數據回答以下問題：

1. **沒有負 reward 的原版模型表現多少？**
2. **掉命給 -1 後，模型最後真正的 Breakout 分數是多少？**
3. **它是不是比較少漏接球？**
4. **它是不是更快開始學會有效策略？**
5. **50 個相同 seed 下，到底有多少局變好、多少局變差？**
6. **改善如果存在，是 final score 的改善，還是只讓模型活得比較久？**
7. **這個 reward shaping 值不值得成為後續預設設定？**

最終結論必須以 **raw Atari game score + paired evaluation** 為主，而不是 shaped training return。
